api/cluster/virtual_service.go

```diff
@@ -0,0 +1,211 @@
+		return nil, err
+	}
+
+	modelVersionRevisionHost := cfg.ModelVersionRevisionURL.Hostname()
+	modelVersionRevisionPath := cfg.ModelVersionRevisionURL.Path
+
+	vs := &v1beta1.VirtualService{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "networking.istio.io/v1beta1",
+			Kind:       "VirtualService",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      cfg.Name,
+			Namespace: cfg.Namespace,
+			Labels:    cfg.Labels,
+		},
+		Spec: istiov1beta1.VirtualService{
+			Gateways: []string{knativeIngressGateway},
+			Hosts:    []string{modelVersionHost},
+			Http:     cfg.createHttpRoutes(modelVersionRevisionHost, modelVersionRevisionPath),
+		},
+	}
+
+	return vs, nil
+}
+
+// getModelVersionHost creates model version endpoint host based on version endpoint's url
+func (cfg *VirtualService) getModelVersionHost() (string, error) {
+	host := strings.Split(cfg.ModelVersionRevisionURL.Hostname(), fmt.Sprintf(".%s.", cfg.Namespace))
+	if len(host) != 2 {
+		return "", fmt.Errorf("invalid version endpoint url: %s. failed to split domain: %+v", cfg.ModelVersionRevisionURL, host)
+	}
+
+	domain := host[1]
+	return fmt.Sprintf("%s-%s.%s.%s", cfg.ModelName, cfg.VersionID, cfg.Namespace, domain), nil
+}
+
+func (cfg *VirtualService) createHttpRoutes(modelVersionRevisionHost, modelVersionRevisionPath string) []*istiov1beta1.HTTPRoute {
+	routeDestinations := []*istiov1beta1.HTTPRouteDestination{
+		{
+			Destination: &istiov1beta1.Destination{
+				Host: defaultIstioIngressGatewayHost,
+			},
+			Headers: &istiov1beta1.Headers{
+				Request: &istiov1beta1.Headers_HeaderOperations{
+					Set: map[string]string{
+						"Host": modelVersionRevisionHost,
+					},
+				},
+			},
+			Weight: 100,
+		},
+	}
+
+	switch cfg.Protocol {
+	case protocol.UpiV1:
+		return []*istiov1beta1.HTTPRoute{
+			{
+				Route: routeDestinations,
+			},
+		}
+
+	default:
+		// Default to application/json if Content-Type header is empty
+		routeDestinationsWithContentType, err := copyRouteDestinations(routeDestinations)
+		if err != nil {
+			log.Errorf("failed to copy routeDestinations: %+v", err)
+			return nil
+		}
+		routeDestinationsWithContentType[0].Headers.Request.Set["Content-Type"] = "application/json"
+
+		uri := &istiov1beta1.StringMatch{
+			MatchType: &istiov1beta1.StringMatch_Exact{
+				Exact: fmt.Sprintf("/v1/models/%s-%s:predict", cfg.ModelName, cfg.VersionID),
+			},
+		}
+		rewrite := &istiov1beta1.HTTPRewrite{
+			Uri: fmt.Sprintf("%s:predict", modelVersionRevisionPath),
+		}
+
+		return []*istiov1beta1.HTTPRoute{
+			{
+				Match: []*istiov1beta1.HTTPMatchRequest{
+					{
+						Uri: uri,
+						Headers: map[string]*istiov1beta1.StringMatch{
+							"content-type": {},
+						},
+					},
+				},
+				Route:   routeDestinationsWithContentType,
+				Rewrite: rewrite,
+			},
+			{
+				Match: []*istiov1beta1.HTTPMatchRequest{
+					{
+						Uri: uri,
+					},
+				},
+				Route:   routeDestinations,
+				Rewrite: rewrite,
+			},
+			{
+				Route: routeDestinations,
```

Comment:
What is this 3rd case supposed to handle (especially in comparison with the 2nd)? Can you share an example?

Reply:
Since standard models are using Kserve V1 inference protocol, this additional routing is for `/v1/models` path (reference: https://kserve.github.io/website/0.11/modelserving/data_plane/v1_protocol/). However, we actually not using API other than Predict API. I'm adding this 3rd case for fallback. But now I'm having second thought, I think we don't really need this one.

Reply:
I see.. No strong opinion against it. If we are keeping it though, let's add a comment. (Actually, thinking about it, it will be better to add an inline comment for each of the cases, to explain what scenario it handles.)